platform/server/elasticsearch/ElasticsearchService.ts

```diff
@@ -77,4 +80,23 @@ export class ElasticsearchService implements CoreService {
   getClusterOfType$(type: ElasticsearchClusterType) {
     return k$(this.clusters$)(map(clusters => clusters[type]));
   }
+
+  getClusterOfType(type: ElasticsearchClusterType): Promise<Cluster> {
+    return k$(this.getClusterOfType$(type))(first(), toPromise());
+  }
+
+  // TODO: implement `getCluster` and `callWithRequest`
+  //
+  // getScopedCluster is not the name I want to use. I really
+  // want `getCluster`, but that exists.
+  // Maybe we supplant that one with this.
+  //
+  // This method is really transparent. It does the two things
+  // we currently do. I think we could work Observables
+  // into this so we get updated stuff.
+  async getScopedCluster(type: ElasticsearchClusterType, headers: Headers): Promise<Cluster> {
+    //const { callWithRequest } = this.getCluster(type);
+    //return (...args) => this.callWithRequest(headers, ...args);
+    return await this.getClusterOfType(type);
```

Comment:
Ah, yeah, not 100% sure what we should do here yet. We can't `new Cluster` because we don't want to create a new elasticsearch-js `Client` per request. For this PR, just to get the `onRequest` change in, maybe we should just return the line above? (the `(...args) => this.callWithRequest(headers, ...args);` line). Then we can do a separate PR where we explore some ideas.

Ah, thinking more about this I think there might be a "fundamental problem" with the current `Cluster` design where we have the two different built-in clients (one for `callWithRequest` and the other for `callWithInternalUser`). The problem is that we might want to implement specific (and different) apis for these two use-cases (when we start building apis like `search` etc on the `Cluster`).

I wonder if we actually want a `getDataCluster` and `getAdminCluster`. Then the `ElasticsearchService` can handle the raw elasticsearch-js clients directly, and the `getDataCluster` would be something like:

```ts
// this creates a new `DataCluster` on every get as they
// need to be scoped to the specified headers
async getDataCluster(headers) {
  const dataClient = await latest(this.dataClient$);
  return new DataCluster(dataClient, headers)
}

// for admin clusters we don't need to scope them,
// so we can create an observable that has the latest
// AdminCluster on it.
async getAdminCluster() {
  return await latest(this.adminCluster$);
}

function latest(obs) {
  // some helper for returning the latest value on obs
}
```

We'd likely want to have observable versions of both too. And maybe rename to something like `getCurrentDataCluster(...)` to make it clearer it won't change over time (because it would be a cluster for the
given config).

Reply:
Also, we don't need a generic `getClusterOfType` style function, I think. I think there's a todo around handling monitoring clusters or something like that in the code, but thinking more about this now, I don't think that's necessary. 

Monitoring currently does this:

```js
req.server.plugins.elasticsearch.getCluster('monitoring');
```

in `master` today, but I don't think that's something we want to do going forward into the new platform. First of, this cluster is only ever used inside monitoring, so we shouldn't make it available everywhere by having monitoring register that with the elasticsearch service. And secondly, if this is needed outside of monitoring that should be exposed by monitoring itself, instead of going through the elasticsearch service.

Otherwise it's causing some strange api designs imo, and it goes against how we want plugins to expose their stuff.

(What we might want to look into at some point though, is perhaps how monitoring can create this cluster. But that's not something we need right now, and I think it should be fairly simple to solve given we solve the data and admin cluster stuff above)

Reply:
We don't currently have tests around elasticsearch service in the new platform, so the code as it is, while it has issues, is passing. we could put a pin in this and do a follow up PR. 

Implementing `callWithCluster` I think will require implementing `callAPI`, `createClient`, and maybe some other lower-level things. Is it worth doing right now?

Reply:
> (What we might want to look into at some point though, is perhaps how monitoring can create this cluster. But that's not something we need right now, and I think it should be fairly simple to solve given we solve the data and admin cluster stuff above)

Security plugin will need its custom `security` cluster as well, but agree it will be easier to solve once we have ready approach for `data` and `admin` clusters.

Reply:
> Security plugin will need its custom `security` cluster as well, but agree it will be easier to solve once we sort out approach for `data` and `admin` clusters.

Is this going to be a new thing, or do we have `security` cluster yet? Are we working with elasticsearch team on that?

Reply:
!@#$!#$ Took me this long, but I finally follow your train of thought, Kim. I had to get to it myself, though. It made no sense until just now. I realized that `Cluster` currently is doing too much, with its two different `Client`s. I like the idea of having Data and Admin Cluster objects that already expose `call` functions so that the consumer doesn't need to care how things happen internally.

I more just wanted to comment about how long I've been staring at this and finally it makes sense.

Reply:
> Is this going to be a new thing, or do we have security cluster yet?

No, it's not a new thing, security plugin currently relies on [`security` cluster](https://github.com/elastic/x-pack-kibana/blob/master/server/lib/get_client_shield.js#L9). So it should just work.